components/gitpod-db/src/typeorm/workspace-db-impl.ts

```diff
@@ -488,35 +488,43 @@ export class TypeORMWorkspaceDBImpl extends TransactionalDBImpl<WorkspaceDB> imp
         limit: number = 100,
         type: WorkspaceType = "regular",
     ): Promise<WorkspaceOwnerAndDeletionEligibility[]> {
-        // we do not allow to run this with a future date
         if (cutOffDate > new Date()) {
             throw new Error("cutOffDate must not be in the future, was: " + cutOffDate.toISOString());
         }
         const workspaceRepo = await this.getWorkspaceRepo();
         const qb = workspaceRepo
             .createQueryBuilder("ws")
-            .select(["ws.id", "ws.ownerId", "ws.deletionEligibilityTime"])
+            .leftJoinAndMapOne(
+                "ws.latestInstance",
+                DBWorkspaceInstance,
+                "wsi",
+                `wsi.id = (
+                SELECT i.id
+                FROM d_b_workspace_instance AS i
+                WHERE i.workspaceId = ws.id
+                ORDER BY i.creationTime DESC
+                LIMIT 1
+            )`,
+            )
+            .select(["ws.id", "ws.ownerId", "ws.deletionEligibilityTime", "wsi.id", "wsi.status"])
             .where("ws.deleted = :deleted", { deleted: 0 })
             .andWhere("ws.type = :type", { type })
             .andWhere("ws.softDeleted IS NULL")
             .andWhere("ws.softDeletedTime = ''")
             .andWhere("ws.pinned = :pinned", { pinned: 0 })
             .andWhere("ws.deletionEligibilityTime != ''")
             .andWhere("ws.deletionEligibilityTime < :cutOffDate", { cutOffDate: cutOffDate.toISOString() })
+            // we don't want to delete workspaces that are active
+            .andWhere(
+                new Brackets((qb) => {
+                    qb.where("wsi.id IS NULL").orWhere("JSON_UNQUOTE(wsi.status->>'$.phase') = :stoppedStatus", {
+                        stoppedStatus: "stopped",
```

Comment:
@filiptronicek will this help us avoid the scenario we saw? For example, it looked like GC triggered workspace stop, and then included the workspace in garbage collection. 

Phrased another way, I'm not sure what criteria is used to stop and delete workspaces when starting garbage collection. Is it here?

Reply:
Here, it looks like we're collecting workspaces to be garbage collected, and only doing on Stopped ones.

Phrased another way, before we start GC, what criteria do we use to stop workspaces, if at all?

Reply:
@kylos101 we should only be stopping workspaces right before actually soft-deleting them. This means that by the time the workspace gc job initialized, they should be running and query-able by that status. 

https://github.com/gitpod-io/gitpod/blob/c6c7c7ba10f1c17b539332198267b84512c045b6/components/server/src/jobs/workspace-gc.ts#L99-L118

https://github.com/gitpod-io/gitpod/blob/c6c7c7ba10f1c17b539332198267b84512c045b6/components/server/src/workspace/workspace-service.ts#L579-L591

Reply:
>  we should only be stopping workspaces right before actually soft-deleting them

It feels like we should never stop a workspace (unless the user requested Stop, the workspace timed out, or the workspace no longer exists in the workspace cluster). 

Maybe I am missing something? I am not sure why we'd stop right before doing soft delete.

Reply:
@kylos101 my assumption is that if we did not stop to-be-soft-deleted workspaces, they would keep on running and incurring costs even though they would be inaccessible to end users. To me, it sounds good that we first clean up by stopping.

With this change, we should **never** be deleting (and hence also not stopping) any running workspaces anymore.

Reply:
With this change, what is the impact to the user? This particular workspace had been running for 9 hours, was stopped, and soft deleted (it wasn't visible in the dashboard). 

It sounds like we'll still stop, but we wont' delete, so it should be visible on the dashboard. Is that right? I assume it would be in the "top level" list of workspaces, and not inactive workspaces?

Reply:
> With this change, what is the impact to the user? This particular workspace had been running for 9 hours, was stopped, and soft deleted (it wasn't visible in the dashboard).

In the case of that workspace, its timeout was set to 24 hours. For some reason, its `deletionEligibilityTime` was wrongly set (this PR sets up logging for us to actually find out why) and was deleted (stopped + soft deleted) when the GC job saw that its `deletionEligibilityTime` was in the past. Not only are we stricter with this change (we do not allow `deletionEligibiltyTime` to be in the past relative to any of its previous values), but we also exclude any non-stopped workspaces.

Both of these new rules should make sure that if we ever encounter the same situation, we would never stop and/or delete running workspaces from the Workspace GC job.

Reply:
Awesome, thanks @filiptronicek !